Add HSM section (#10)
NLnetLabs/krill-manual: source/hsm.rst

.. _doc_krill_hsm:

Hardware Signing Modules
========================

> Krill uses openssl by default for key generation, and it stores the

Reviewer: s/openssl/OpenSSL/

Author: check

private keys thus generated in the `keys` directory under its data
directory. For many organisations this will be good enough, provided
of course that they ensure that access to the server and backup data
is restricted.

If you want to take security one (big) step up, then you may want to

> use a Hardware Signing Module (HSM) instead. HSMs are devices that can

Reviewer: Isn’t it Hardware Security Module?

be used to safeguard and manage digital keys. They are designed to allow
applications to use private keys, whilst ensuring that the actual
private keys are never leaked outside of the device - not even to the
application that is using the key. The use of HSMs is supported in

> Krill 0.10.0-rc1 and up.

Reviewer 2: Referring to RC1 is going to be outdated in a week. Why not just refer to Krill 0.10.0?

Author: Well... technically it won't be outdated - it would still be true albeit less relevant perhaps when this is tagged to the actual stable release for 0.10.0. On the other hand: saying 0.10.0 when it's not out yet could potentially be confusing to users wanting to test rc1.

I do see what you mean, but I was indeed planning to just update it when the 0.10.0 release happens.

Krill uses what it calls a "signer" to create and manage keys and to
sign data with them. For the most part the Krill CA code is unaware of
which signer implementation is associated with a key. For long-lived
key-pairs such as the keys used in RPKI CA certificates (the resource
certificate signed by the parent) and the ID certifcate used in the
:RFC:`6492` (provisioning) and :RFC:`8181` (publication) protocols, it
will only keep track of the public key identifier.

It falls to the "signer" then to map these public key identifier to an
actual private key that can be used for signing operations. As mentioned,
the default signer uses OpenSSL, in which case the actual private keys
are simply stored on disk in the `keys` sub-directory of Krill data
directory.

If you have access to a Hardware Security Module (HSM) you can instead
configure "signer" implementations which will use the HSM to create and
safeguard the private keys and perform any signing operation inside the
HSM.

.. Note:: Krill uses one-off signing keys for the EE certificates used
          in RPKI Signed Objects (such as ROAs and Manifests). These
          keys are generated whenever such an object is created, and

>           used only once for signing, and then they are detroyed.

Reviewer 2: ```suggestion
          used only once for signing, and then they are destroyed.
```

.. _#732: https://github.com/NLnetLabs/krill/issues/732

In order to work with Krill the HSM must support the following operations:

===================  =================
PKCS#11              KMIP
===================  =================
C_CloseSession       Activate
C_DeleteObject       Create Key Pair
C_Finalize           Destroy
C_FindObjects        Get
C_FindObjectsFinal   Modify Attribute
C_FindObjectsInit    Query
C_GenerateKeyPair    Revoke
C_GetAttributeValue  Sign
C_GetInfo
C_GetSlotInfo
C_GetSlotList
C_GetTokenInfo
C_Initialize
C_Login
C_OpenSession
C_Sign
C_SignInit
===================  =================

Krill can use a cluster of HSMs if the cluster appears to Krill as a
single HSM, i.e. if Krill is not aware that the "single" HSM is in fact
a cluster of HSMs.

PKCS#11 or KMIP?
""""""""""""""""

PKCS#11 and KMIP are very similar in the capabilities they provide, so
much so that there are commercial offerings that can bridge from one to
the other, HSMs may offer support for both and both standards are
maintained by `OASIS <https://www.oasis-open.org/>`_. From a Krill server
operation perspective however they are very different and each has its

> own PROs and CONs.

Reviewer 2: ```suggestion
own pros and cons.
```
https://www.merriam-webster.com/dictionary/pros%20and%20cons

PKCS#11 works by delegating configuration, logging, administration,
maintenance and upgrade of the interface with the HSM to a library file
outside of Krill that Krill loads when it runs. You therefore have to
manage and monitor this library and its logs as a separate component on
the system running Krill. However, as a separate component it can
connect in any way it needs to the backend which can be local or remote,
or possibly even to a cluster of systems. Krill sees only the library,
it has no way of knowing whether the backend is local or remote,

> singlular or clustered. However it also has no way of controlling how

Reviewer 2: ```suggestion
singular or clustered. This means it also has no way of controlling how
```
Typo, and avoid repeating 'however'.

long the library will block to wait for a task to complete or how many
requests it can handle at once or how many system resources it uses.

KMIP is arguably simpler to setup. With KMIP you only need to manage
Krill and the HSM, there is no additional library component to manage as
with PKCS#11. Krill itself communicates directly with the HSM and so all

> configuration, logging & resource usage is determined by Krill and

Reviewer 2: ```suggestion
configuration, logging and resource usage is determined by Krill. In addition, 
```

monitoring is done by monitoring Krill itself. Krill connects to the
KMIP server via TLS encrypted TCP and thus could also potentially be
routed to one of many backend servers in a cluster, or the server could
be a process running locally on the same host such as PyKMIP.

Scenarios
---------

Fresh installation
""""""""""""""""""

With a fresh installation of Krill you can use the HSM from the start.
No keys will be stored locally, instead all long-lived keys will be
stored in the HSM.

Migrating to or between HSMs
""""""""""""""""""""""""""""

Krill does not support migration of existing RPKI CA private keys from
one signer to another. Instead you will need to perform a

> :ref:`key rollover<doc_krill_ca_keyroll>` for each CA. **NOTE:** Not all
> keys can be rolled. See the warning above about migration of ID keys
> used in parent/child and CA/publication server relationships.

Reviewer 2: ```suggestion
:ref:`key rollover<doc_krill_ca_keyroll>` for each CA.

.. Note:: Not all keys can be rolled. See the warning above about
          migration of ID keys used in parent/child and CA/publication 
          server relationships.
 
```

Signer Roles
""""""""""""

When configuring more than one signer, one may be designated the
``default_signer`` and another (or the same one) may be designated the
``one_off_signer``. The ``default_signer`` is used to create all new
keys, except in the case of one-off signing for which the
``one_off_signer`` signer will be used to create a new temporary key,
sign with it then destroy it.

Specifying the ``default_signer`` and ``one_off_signer`` is done by
referencing the name of the signer. For example the above is equivalent
to:

.. code-block::

   default_signer = "Default OpenSSL signer"
   one_off_signer = "Default OpenSSL signer"

   [[signers]]
   type = "OpenSSL"
   name = "Default OpenSSL signer"

When only a single signer is defined it will implicitly be the
``default_signer``. When defining more than one signer the
``default_signer`` must be set explicitly.

If the ``default_signer`` is not of type ``OpenSSL`` and is not
explicitly set as the ``one_off_signer``, an OpenSSL signer will
automatically be used as the ``one_off_signer``.

Configuring a PKCS#11 signer
""""""""""""""""""""""""""""

.. Note:: To actually use a PKCS#11 based signer you must first set it

>           up according to the vendors instructions. This may require

Reviewer 2: ```suggestion
          up according to the vendor's instructions. This may require
```

file for the HSM that was supplied by the HSM provider and a slot ID or
label, and if needed, a user pin.

.. code-block::

   [[signers]]
   type = "PKCS#11"
   name = "SoftHSMv2 via PKCS#11"
   lib_path = "/usr/local/lib/softhsm/libsofthsm2.so"
   slot = 0x12a9f8f7
   user_pin = "xxxx"                                       # optional
   login = true                                            # optional, default = true

Note:
  - If using a slot label rather than ID you can supply the label using ``slot = "my label"``.
  - You can also supply an integer slot ID, e.g. ``slot = 123456``.
  - If your HSM does not require you to login you can set ``login = false``.
  - If your HSM requires you to supply a pin via an external key pad you can omit the ``user_pin`` setting.

Configuring a KMIP signer
"""""""""""""""""""""""""

.. note:: To actually use a KMIP based signer you must first set it up
          according to the vendors instructions. This may require
          setting up users and passwords and/or obtaining certificates
          in order to populate the associated settings in the
          ``krill.conf`` file.

> For a KMIP signer you must specify the host FQDN or IP address, and

Reviewer 2: ```suggestion
For a KMIP signer you must specify the fully-qualified domain name (FQDN) or IP address of the host, and
```

             will be unable to sign with that key!

             One particular scenario to watch out for is when
             reconfiguring an existing Krill instance to use an HSM when
             that Krill instance already has at least one CA (and thus
             already created at least one key pair using OpenSSL).

             In this scenario, if the changes to ``krill.conf`` to use
             the HSM define only the one signer (the HSM) and do NOT set
             that signer as the ``one_off_signer``, then Krill will
             activate the default OpenSSL signer for one-off key signing
             and will use it to find the previously created OpenSSL keys.

             If however the one and only HSM signer is also set as the
             ``one_off_signer`` then Krill will not activate the OpenSSL
             signer and so will not find the previously created OpenSSL
             keys. In this case you must explicitly add a ``[[signers]]``
             block of ``type = "OpenSSL"`` with default settings thereby
             causing Krill to activate the default OpenSSL signer.

SoftHSMv2 Example
-----------------

> Lets see how to setup `SoftHSMv2 <https://github.com/opendnssec/SoftHSMv2>`_

Reviewer 2: ```suggestion
Let's see how to setup `SoftHSMv2 <https://github.com/opendnssec/SoftHSMv2>`_
```
